feat(bundler): Switch to `ast-grep` parser

zharinov opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

### Gemfile Parser

A comprehensive Ruby Gemfile parser for Renovate built with AST-grep. The parser is organized into focused modules:

- **`index.ts`** - `parseGemfile()` top-level function
- **`common.ts`** - Shared AST utilities and Ruby language loading, i.g. variable resolution
- **`gem.ts`** - Gem name/version parsing (`gem "rails", "~> 7.0"`)
- **`git.ts`** - Git dependency handling (`git:`, `github:`, `ref:`, `tag:`, `branch:`)
- **`group-scope.ts`** - Dependency groups and scoping (`group :development do ... end`)
- **`ruby-version.ts`** - Ruby version requirements (`ruby "2.7.1"`)
- **`source-scope.ts`** - Registry/source URL management (`source "https://..." do ... end`)

Handles complex Gemfile syntax including nested groups, source blocks, variable resolution, version constraints, and edge cases.

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

zharinov committed:
feat(bundler): Switch to `ast-grep` parser

zharinov committed:
Refactor

zharinov committed:
Fixes

zharinov committed:
Variables

zharinov committed:
Fix signature

zharinov committed:
Fix tests

zharinov committed:
Remove local package tests from bundler parser

zharinov committed:
Merge branch 'main' into feat/bundler-ast-grep-parser

zharinov committed:
Fix lockfile

zharinov committed:
Fix line numbers

zharinov committed:
Fix snapshots

zharinov committed:
Fix autoreplace

zharinov commented:
`e2e` and `docker` build failures have something to do with `postinstall` script whitelisting strategy of `pnpm`

viceice commented on package.json:
```diff
@@ -369,6 +372,7 @@
       "safe-json-stringify"
     ],
     "onlyBuiltDependencies": [
+      "@ast-grep/lang-ruby",
```
i think this needs to be added to this file too
https://github.com/renovatebot/renovate/blob/main/test/e2e/package.json#L9

zharinov committed:
Fix CI

zharinov committed:
Refactor

zharinov committed:
Replace GitRefData interface with type from PackageDependency

zharinov committed:
Rename bundler Ruby version parser files

zharinov committed:
Refactor

zharinov committed:
Fix coverage

zharinov committed:
Refactor

zharinov committed:
Update git.spec.ts

zharinov committed:
Refactor

zharinov committed:
Add support for boolean key-value args in Bundler parser

zharinov committed:
Add support for optional gem groups

The commit focuses on adding proper handling of `optional` groups for
Ruby gems in the Bundler manager.

zharinov committed:
Add support for double Ruby version range extraction

zharinov committed:
Merge branch 'main' into feat/bundler-ast-grep-parser

zharinov committed:
Merge branch 'main' into feat/bundler-ast-grep-parser

zharinov committed:
Merge branch 'main' into feat/bundler-ast-grep-parser

secustor commented on lib/modules/manager/bundler/extract.spec.ts:
```diff
@@ -267,17 +305,24 @@ describe('modules/manager/bundler/extract', () => {
+          datasource: 'rubygems',
           depName: 'gem_with_one_value',
-          currentValue: '">= 3.0.5"',
+          skipReason: 'unknown-registry',
```
What triggered this change?

zharinov replied:
In Gemfile, at least one registry is mandatory, but anyways I reverted back to the original Renovate as of `main` at the moment.

zharinov committed:
Remove skipReason from Bundler parser and extract tests

zharinov committed:
Bump @ast-grep/napi from 0.39.1 to 0.39.3

zharinov committed:
Remove unused common-tags import

viceice reviewed (commented):
otherwise LGTM

rarkins reviewed (commented):
Let's merge after v42

zharinov committed:
Merge branch 'main' into feat/bundler-ast-grep-parser

zharinov committed:
Merge branch 'main' into feat/bundler-ast-grep-parser

zharinov committed:
Upgrade ast-grep and tree-sitter packages

zharinov committed:
Refactor ast-grep anchor node handling

Fix bug in ast-grep pattern matching by moving anchor node check from
bundler parser into the core matching function.

zharinov committed:
Add ast-grep ROOT_NODE match comment for clarity
